PartitionMigrationListenerTest.testMigrationStats_whenMigrationProcessCompletes

_master_ (commit 58abe5433401a6d857c23710cfc9813dde0eb3d2)

Failed on Windows : http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-Windows-OracleJDK8/120/testReport/com.hazelcast.partition/PartitionMigrationListenerTest/testMigrationStats_whenMigrationProcessCompletes/

**Stacktrace**
```
java.lang.AssertionError: expected:<271> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.partition.PartitionMigrationListenerTest.assertMigrationProcessEventsConsistent(PartitionMigrationListenerTest.java:201)
	at com.hazelcast.partition.PartitionMigrationListenerTest.testMigrationStats_whenMigrationProcessCompletes(PartitionMigrationListenerTest.java:83)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/test/java/com/hazelcast/partition/PartitionMigrationListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/partition/PartitionMigrationListenerTest.java
@@ -18,12 +18,10 @@ package com.hazelcast.partition;
 
 import com.hazelcast.cluster.ClusterState;
 import com.hazelcast.config.Config;
-import com.hazelcast.config.ListenerConfig;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.internal.partition.MigrationInfo;
 import com.hazelcast.internal.partition.MigrationStateImpl;
 import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
-import com.hazelcast.internal.partition.impl.MigrationCommitTest.DelayMigrationStart;
 import com.hazelcast.internal.partition.impl.MigrationInterceptor;
 import com.hazelcast.internal.partition.impl.MigrationStats;
 import com.hazelcast.internal.util.UuidUtil;
@@ -43,7 +41,6 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.UUID;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
@@ -73,10 +70,16 @@ public class PartitionMigrationListenerTest extends HazelcastTestSupport {
         HazelcastInstance hz1 = factory.newHazelcastInstance();
         warmUpPartitions(hz1);
 
+        // Change to NO_MIGRATION to prevent repartitioning
+        // before 2nd member started and ready.
+        hz1.getCluster().changeClusterState(ClusterState.NO_MIGRATION);
+
         EventCollectingMigrationListener listener = new EventCollectingMigrationListener();
         hz1.getPartitionService().addMigrationListener(listener);
 
-        factory.newHazelcastInstance();
+        HazelcastInstance hz2 = factory.newHazelcastInstance();
+        // Back to ACTIVE
+        changeClusterStateEventually(hz2, ClusterState.ACTIVE);
 
         MigrationEventsPack eventsPack = listener.ensureAndGetSingleEventPack();
         assertMigrationProcessCompleted(eventsPack);
@@ -91,10 +94,10 @@ public class PartitionMigrationListenerTest extends HazelcastTestSupport {
         int partitionCount = 100;
         config.setProperty(ClusterProperty.PARTITION_COUNT.getName(), String.valueOf(partitionCount));
 
-        HazelcastInstance hz = factory.newHazelcastInstance(config);
-        warmUpPartitions(hz);
+        HazelcastInstance hz1 = factory.newHazelcastInstance(config);
+        warmUpPartitions(hz1);
 
-        InternalPartitionServiceImpl partitionService = (InternalPartitionServiceImpl) getPartitionService(hz);
+        InternalPartitionServiceImpl partitionService = (InternalPartitionServiceImpl) getPartitionService(hz1);
         AtomicReference<HazelcastInstance> newInstanceRef = new AtomicReference<>();
         partitionService.setMigrationInterceptor(new MigrationInterceptor() {
             @Override
@@ -111,13 +114,20 @@ public class PartitionMigrationListenerTest extends HazelcastTestSupport {
         });
 
         EventCollectingMigrationListener listener = new EventCollectingMigrationListener();
-        hz.getPartitionService().addMigrationListener(listener);
+        hz1.getPartitionService().addMigrationListener(listener);
+
+        // Change to NO_MIGRATION to prevent repartitioning
+        // before 2nd member started and ready.
+        hz1.getCluster().changeClusterState(ClusterState.NO_MIGRATION);
 
         // trigger migrations
-        factory.newHazelcastInstance(config);
+        HazelcastInstance hz2 = factory.newHazelcastInstance(config);
+
+        // Back to ACTIVE
+        changeClusterStateEventually(hz2, ClusterState.ACTIVE);
 
         // await until 3rd member joins
-        assertClusterSizeEventually(3, hz);
+        assertClusterSizeEventually(3, hz1);
         assertTrueEventually(() -> assertNotNull(newInstanceRef.get()));
 
         List<MigrationEventsPack> eventsPackList = listener.ensureAndGetEventPacks(2);
@@ -132,6 +142,16 @@ public class PartitionMigrationListenerTest extends HazelcastTestSupport {
 
         // 2nd migration process finishes by consuming all migration tasks
         MigrationEventsPack secondEventsPack = eventsPackList.get(1);
+
+        if (secondEventsPack.migrationProcessCompleted.getCompletedMigrations() == 1
+            && !secondEventsPack.migrationsCompleted.get(0).isSuccess()) {
+            // There is a failed migration process
+            // because migrations restarted before 3rd member is ready.
+            // This migration process is failed immediately
+            // and we expect a third migration process.
+            secondEventsPack = listener.ensureAndGetEventPacks(3).get(2);
+        }
+
         assertMigrationProcessCompleted(secondEventsPack);
         assertMigrationProcessEventsConsistent(secondEventsPack);
         assertMigrationEventsConsistentWithResult(secondEventsPack);
@@ -237,19 +257,19 @@ public class PartitionMigrationListenerTest extends HazelcastTestSupport {
         int partitionCount = 10;
         config.setProperty(ClusterProperty.PARTITION_COUNT.getName(), String.valueOf(partitionCount));
 
-        // hold the migrations until all nodes join so that there will be no retries / failed migrations etc.
-        CountDownLatch migrationStartLatch = new CountDownLatch(1);
-        config.addListenerConfig(new ListenerConfig(new DelayMigrationStart(migrationStartLatch)));
-
         HazelcastInstance instance1 = factory.newHazelcastInstance(config);
         warmUpPartitions(instance1);
+
+        // Change to NO_MIGRATION to prevent repartitioning
+        // before 2nd member started and ready.
+        instance1.getCluster().changeClusterState(ClusterState.NO_MIGRATION);
 
         CountingMigrationListener migrationListener = new CountingMigrationListener(partitionCount);
         instance1.getPartitionService().addMigrationListener(migrationListener);
 
         HazelcastInstance instance2 = factory.newHazelcastInstance(config);
 
-        migrationStartLatch.countDown();
+        changeClusterStateEventually(instance2, ClusterState.ACTIVE);
 
         waitAllForSafeState(instance2, instance1);
 

